### PR TITLE
chore: Omit analyticsSpecs without instanceSize in INFINITE PATCH payloads

### PR DESCRIPTION
**Pull request by lantoli**
## Description

When clearing `storage_config` on an INFINITE cluster, the provider was sending an invalid PATCH payload that Atlas rejected with 400 MISSING_ATTRIBUTE. The payload included `analyticsSpecs` with `nodeCount=0` but no `instanceSize`.

### Root Cause

When clearing `storage_config`, the provider forces `replicationSpecs` into the PATCH. The plan modifier didn't copy `analytics_specs` from state when `node_count == 0` (to avoid ANALYTICS_INSTANCE_SIZE_MUST_MATCH errors), so `CopyUnknowns` produced `analyticsSpecs: {"nodeCount": 0}` without `instanceSize`. Atlas requires `instanceSize` when `replicationSpecs` is present.

### The Fix

1. **Explicit null for storage removal**: Use `SetStorageConfigNil()` to explicitly set `storageConfig` to null in the plan. This creates a detectable change (replacement, not removal) that `PatchPayload` recognizes, ensuring `replicationSpecs` is included in the PATCH. Atlas clears the limit because `storageConfig` is absent.

2. **Preserve analytics intent during planning**: When the plan has `analytics_specs` with `node_count = 0` but no `instance_size` (explicit removal), copy `instance_size` from state to create a valid removal request.

3. **Keep request cleanup**: `omitInvalidInfiniteConfig` remains necessary to remove empty auto-scaling children and any remaining invalid `analyticsSpecs`.

### Why Analytics Needs Special Handling (But Read-Only Doesn't)

Read-only removal always works because the plan modifier has a fallback chain that ensures `instance_size` is present: state → electable specs → region electable specs. Analytics removal needed special handling because the original code avoided copying from state to prevent ANALYTICS_INSTANCE_SIZE_MUST_MATCH errors during scaling operations. The fix preserves `instance_size` from state when the user explicitly sets `node_count = 0` (removal intent).

### Key Changes

- **Replaced `ForceUpdateAttr` with explicit null**: More targeted approach that works with Atlas's replacement semantics
- **Simplified storage-removal detection**: `setStorageConfigNil` returns whether it marked any removals, eliminating the need for a separate scan
- **Fixed analytics removal edge case**: Plan modifier now preserves `instance_size` for explicit removals
- **Fixed plural data source checks**: Replaced hardcoded `results.0` indices with dynamic lookup by name in effective fields tests, preventing failures when INFINITE clusters (which don't have disk_size_gb) are returned

Link to any related issue(s): CLOUDP-447210

## Type of change:

- [x] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fix and verified my code
- [x] If changes include deprecations or removals I have added appropriate changelog entries.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

## Further comments

### CODE REVIEW

File: internal/service/advancedcluster/effective_fields_test.go [hunk 1]
```diff
@@ -671,28 +685,43 @@ func (req effectiveReq) check() resource.TestCheckFunc {
...
+			project_id   = %[1]q
+			name         = %[2]q
+			cluster_type = "REPLICASET"
+
+			replication_specs = [{
+				region_configs = [{
+					provider_name = %[3]q
+					region_name   = "US_EAST_1"
+					priority      = 7
+					electable_specs = {
```
[the remark this answers is not in the document]
Comment (lantoli) [reply internal/service/advancedcluster/effective_fields_test.go]: You're right, thanks for catching this. I've reverted the change to configEffectiveTenantFlex to not include electable_specs, so the test can still verify the error case where electableSpecs is not specified.

File: internal/service/advancedcluster/resource_database_edition_test.go [hunk 2]
```diff
@@ -353,6 +505,47 @@ func configDatabaseEditionWithDiskGBEnabled(projectID, clusterName string, datab
...
+			pit_enabled    = true
+
+			replication_specs = [{
+				region_configs = [{
+					electable_specs = {
+						instance_size = "M10"
+						node_count    = 2
+					}
+					analytics_specs = {
+						instance_size = "M10"
```
[the remark this answers is not in the document]
Comment (lantoli) [reply internal/service/advancedcluster/resource_database_edition_test.go]: Good catch. I've updated configDatabaseEditionWithZeroNodeAnalytics to omit instance_size when node_count = 0, so the test now actually reproduces the invalid payload scenario that prompted the fix.

File: internal/service/advancedcluster/effective_fields_test.go [hunk 3]
```diff
@@ -544,10 +567,9 @@ func (req effectiveReq) config() string {
 
 func (req effectiveReq) check() resource.TestCheckFunc {
 	const (
-		specsPath           = "replication_specs.0.region_configs.0.electable_specs."
-		effectivePath       = "replication_specs.0.region_configs.0.effective_electable_specs."
-		effectivePathPlural = "results.0.replication_specs.0.region_configs.0.effective_electable_specs."
-		autoScalingPath     = "replication_specs.0.region_configs.0.auto_scaling."
+		specsPath       = "replication_specs.0.region_configs.0.electable_specs."
+		effectivePath   = "replication_specs.0.region_configs.0.effective_electable_specs."
+		autoScalingPath = "replication_specs.0.region_configs.0.auto_scaling."
```
[the remark this answers is not in the document]
Comment (lantoli) [reply internal/service/advancedcluster/effective_fields_test.go:577]: You're right, I've added pluralEffectiveSpecsCheck to all tests that use the plural data source, including effectiveBasic, effectiveUnsetToSet, and effectiveSetToUnset.

File: internal/service/advancedcluster/effective_fields_test.go [hunk 4]
```diff
@@ -671,28 +685,43 @@ func (req effectiveReq) check() resource.TestCheckFunc {
...
+
+			replication_specs = [{
+				region_configs = [{
+					provider_name = %[3]q
+					region_name   = "US_EAST_1"
+					priority      = 7
+					electable_specs = {
+						instance_size = "M10"
+						node_count    = 3
 					}
```
[the remark this answers is not in the document]
Comment (lantoli) [reply internal/service/advancedcluster/effective_fields_test.go]: You're right, I've reverted the change to configEffectiveTenantFlex to not include electable_specs, so the test can still verify the error case where electableSpecs is not specified.

File: internal/service/advancedcluster/plan_modifier.go [hunk 5]
```diff
@@ -135,6 +137,18 @@ func adjustRegionConfigsChildren(ctx context.Context, diags *diag.Diagnostics, s
 					return
 				}
 				planRegionConfigsTF[j].AnalyticsSpecs = objType
+			} else if planAnalyticsSpecs != nil && planAnalyticsSpecs.NodeCount.ValueInt64() == 0 && (planAnalyticsSpecs.InstanceSize.IsNull() || planAnalyticsSpecs.InstanceSize.IsUnknown()) {
```
[the remark this answers is not in the document]
Comment (lantoli) [reply internal/service/advancedcluster/plan_modifier.go]: Good point. I've updated the condition to explicitly check that node_count is not null and not unknown before checking if it's 0. This distinguishes between explicit node_count = 0 and unresolved/omitted values.

File: internal/service/advancedcluster/resource_autoscaling_test.go [hunk 6]
```diff
@@ -212,6 +222,74 @@ func TestUpdateRemovesShardSizeLimitPreservesPlannedValues(t *testing.T) {
...
+		require.Len(t, payload.GetReplicationSpecs(), 1)
+		encoded, err := json.Marshal(payload.GetReplicationSpecs()[0].GetRegionConfigs())
+		require.NoError(t, err)
+		t.Logf("PATCH request: %s", string(encoded))
+
+		// The request should include analyticsSpecs with nodeCount=0 and instanceSize
+		// to properly remove the analytics nodes
+		require.Contains(t, string(encoded), `"analyticsSpecs"`, "PATCH should include analyticsSpecs for removal")
+		require.Contains(t, string(encoded), `"nodeCount":0`, "analyticsSpecs should have nodeCount=0")
+		require.Contains(t, string(encoded), `"instanceSize"`, "analyticsSpecs should have instanceSize")
```
[the remark this answers is not in the document]
Comment (lantoli) [reply internal/service/advancedcluster/resource_autoscaling_test.go]: You're right. I've updated the test to decode the analyticsSpecs object directly and assert its own instanceSize, rather than searching the entire region JSON.

File: internal/service/advancedcluster/resource_database_edition_test.go [hunk 7]
```diff
@@ -353,6 +505,47 @@ func configDatabaseEditionWithDiskGBEnabled(projectID, clusterName string, datab
...
+						node_count    = 0
```
[the remark this answers is not in the document]
Comment (lantoli) [reply internal/service/advancedcluster/resource_database_edition_test.go:534]: You're right. I've updated configDatabaseEditionWithZeroNodeAnalytics to omit instance_size when node_count = 0, so the test now actually reproduces the invalid payload scenario that prompted the fix.

File: internal/service/advancedcluster/common.go [hunk 8]
```diff
@@ -57,33 +57,51 @@ func AddIDsToReplicationSpecs(replicationSpecs []admin.ReplicationSpec20240805,
...
 
-func regionHasShardSizeLimit(regionConfig *admin.CloudRegionConfig20240805) bool {
-	return regionConfig.AutoScaling != nil && regionConfig.AutoScaling.StorageConfig != nil && regionConfig.AutoScaling.StorageConfig.HasShardSizeLimitGB()
-}
-
-func omitEmptyAutoScalingChildren(replicationSpecs []admin.ReplicationSpec20240805) {
+// omitInvalidInfiniteConfig removes empty auto-scaling children and analyticsSpecs without
+// instanceSize that Atlas rejects. This is needed because conversion can produce invalid request objects
+// when the entire replicationSpecs is included in the PATCH.
+func omitInvalidInfiniteConfig(replicationSpecs []admin.ReplicationSpec20240805) {
```
Comment (EspenAlbert) [inline internal/service/advancedcluster/common.go:94]: q: We have already changed the plan modifier in this PR, why do we also have to change the update here?

Comment (lantoli) [reply internal/service/advancedcluster/common.go:94]: they don't overlap, they work on different representations. The plan modifier runs on TF values before apply (it copies `instance_size` from state so the plan stays valid on analytics removal). `omitInvalidInfiniteConfig` runs on the SDK request after conversion, where the converter can still emit shapes Atlas rejects like empty `compute`/`diskGB` or `analyticsSpecs` with `nodeCount: 0` and no `instanceSize`, since `replicationSpecs` is atomic and any change re-sends all regions.

It's needed now because this PR removes `ForceUpdateAttr = ["replicationSpecs"]`. The forced full-list PATCH used to mask these artifacts; with the request built from the actual diff they surface and need explicit handling. The plan modifier only covers the specific removal case and doesn't run on Create.

In a follow-up I plan to remove the `ForceUpdateAttr` logic altogether to simplify the update flow

Comment (EspenAlbert) [reply internal/service/advancedcluster/common.go:94]: Discussed offline. Thanks for the explanation

**Review by EspenAlbert — APPROVED** [review]
Thanks for the discussion!
